content/advent-2018/postmortem-debugging-delve.md

```diff
@@ -0,0 +1,469 @@
+        "optimized": true
+      }
+    },
+    "threadID": 0
+  },
+  ...
+```
+
+Every object in the JSON represents a single goroutine. By checking [“goroutines” command help](https://github.com/derekparker/delve/blob/master/Documentation/cli/README.md#goroutines) we can figure out what data Delve knows. We're interested in `userCurrentLoc` field, which is "topmost stackframe in user code".
+
+Of all goroutines in the JSON let's list unique function names with the exact line numbers where function got paused:
+
+```
+= jq -c '.result[] | [.userCurrentLoc.function.name, .userCurrentLoc.line]' server-test-1_dlv-rpc-list_goroutines.json | sort | uniq -c
+
+   1 ["internal/poll.runtime_pollWait",173]
+1000 ["main.(*Metrics).CountS",113]
+   1 ["main.(*Metrics).SetM",145]
+   1 ["main.(*Metrics).startOutChannelConsumer",239]
+   1 ["main.run",56]
+   1 ["os/signal.signal_recv",139]
+   6 ["runtime.gopark",303]
+```
+
+The majority of goroutines (1000) have stacked in `main.(*Metrics).CountS:113`. Now, this is the perfect time to look at the source code.
+
+In the `main` package, find `Metrics` struct and look at its `CountS` method (see `example/server/metrics.go`):
+
+```go
+// CountS increments counter per second.
+func (m *Metrics) CountS(key string) {
+    // ···
+
+    m.inChannel <- NewCountMetric(key, 1, second)
+}
+```
+
+Our server has stacked on sending to the `inChannel` channel. Let’s find out who is supposed to read from this channel. After inspecting the code, we should find the following function (example/server/metrics.go):
+
+```
+// starts a consumer for inChannel
+func (m *Metrics) startInChannelConsumer() {
+    for inMetrics := range m.inChannel {
+   	    // ···
+    }
+}
+```
+
+The function reads values out of the channel and does something with them, one by one. In what possible situations, could the sending to this channel being blocked?
+
+When working with channels, there are only four possible "oopsies", according to Dave Cheney's [Channel Axioms](https://dave.cheney.net/2014/03/19/channel-axioms):
+
+- send to a nil channel block forever
+- receive from a nil channel block forever
+- send to a closed channel panics
+- receive from a closed channel returns the zero value immediately.
+
+"Send to a nil channel block forever" – at first sight, this sounds like something possible. But, after double-checking with the code, `inChannel` is initialised in the `Metrics` constructor. So it can't be nil.
+
+Let’s look at the list of functions we’ve got above. Could this (buffered) channel become full because we've stacked somewhere in `(*Metrics).startInChannelConsumer()`?
+
+As you may notice, there is no `startInChannelConsumer` method in the list at all. But what if we stacked somewhere below the method’s callstack?
+
+Delve provides the start position from where we came to the user location, that is `startLoc` field in the JSON. Search for goroutines whose start location was in `startInChannelConsumer` function:
```

Comment:
I think it'd be good to explain "user location" versus "runtime location" so that it becomes clear what is meant here. Also, in the paragraph above: "below the method's callstack" is hard to understand for readers without a lot of experience with debuggers.